test_rmw_implementation/test/test_subscription.cpp

```diff
@@ -524,6 +524,79 @@ TEST_F(CLASSNAME(TestSubscriptionUse, RMW_IMPLEMENTATION), take_with_info_with_b
   sub->implementation_identifier = implementation_identifier;
 }
 
+TEST_F(CLASSNAME(TestSubscriptionUse, RMW_IMPLEMENTATION), ignore_local_publications) {
+  rmw_ret_t ret;
+  bool taken = false;
+
+  // Create publisher
+  rmw_publisher_options_t pub_options = rmw_get_default_publisher_options();
+  rmw_publisher_t * pub = rmw_create_publisher(node, ts, topic_name, &qos_profile, &pub_options);
+  ASSERT_NE(nullptr, pub) << rmw_get_error_string().str;
+
+  // Create subscription with ignore_local_publications = true
+  rmw_subscription_options_t sub_options_ignorelocal = rmw_get_default_subscription_options();
+  sub_options_ignorelocal.ignore_local_publications = true;
+  rmw_subscription_t * sub_ignorelocal =
+    rmw_create_subscription(node, ts, topic_name, &qos_profile, &sub_options_ignorelocal);
+  ASSERT_NE(nullptr, sub_ignorelocal) << rmw_get_error_string().str;
+
+  size_t subscription_count = 0u;
+  SLEEP_AND_RETRY_UNTIL(rmw_intraprocess_discovery_delay, rmw_intraprocess_discovery_delay * 10) {
+    ret = rmw_publisher_count_matched_subscriptions(pub, &subscription_count);
+    if (RMW_RET_OK == ret && 2u == subscription_count) {  // Early return on failure.
+      break;
+    }
+  }
+
+  // Roundtrip message from publisher to both subscriptions
+  test_msgs__msg__BasicTypes original_message{};
+  ASSERT_TRUE(test_msgs__msg__BasicTypes__init(&original_message));
+  original_message.bool_value = true;
```

Comment:
@jamoralp consider using `OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT` as in line 750 to ensure message destruction on scope exit.